### PR TITLE
Events in grouped object

### PR DESCRIPTION
**Pull request by asturur**
Preliminary cleaning and test.

First i remove some duplicate calls of getPointer(e, true); We are passing event e, and in every function searching for the pointer from the event. While we can search pointer once and then just pass reference to it.

Then a nice logic to discern when we should go deep in the groups and when we should stop at first object is still needed.

### CODE REVIEW

**Comment by asturur** [comment]
it detects the object in groups and nested groups, but the mouse over and mouse out logic needs tweaks. i think we need to return an array of found target instead of a single target. The first one gets normal processing (selection, grouping...) , the others just fires mouse up, mouse down, mouse over, mouse out...

**Comment by asturur** [comment]
needed for testing
```
{"objects":[{"type":"group","originX":"left","originY":"top","left":160.02,"top":107,"width":216.49,"height":262,"fill":"rgb(0,0,0)","stroke":null,"strokeWidth":0,"strokeDashArray":null,"strokeLineCap":"butt","strokeLineJoin":"miter","strokeMiterLimit":10,"scaleX":1,"scaleY":1,"angle":0,"flipX":false,"flipY":false,"opacity":1,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"objects":[{"type":"circle","originX":"left","originY":"top","left":-62.27,"top":-31,"width":80,"height":80,"fill":"rgb(0,0,255)","stroke":null,"strokeWidth":1,"strokeDashArray":null,"strokeLineCap":"butt","strokeLineJoin":"miter","strokeMiterLimit":10,"scaleX":1,"scaleY":2,"angle":0,"flipX":false,"flipY":false,"opacity":0.5,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"radius":40,"startAngle":0,"endAngle":6.283185307179586},{"type":"rect","originX":"left","originY":"top","left":-62.27,"top":-131,"width":50,"height":50,"fill":"rgb(255,0,0)","stroke":null,"strokeWidth":1,"strokeDashArray":null,"strokeLineCap":"butt","strokeLineJoin":"miter","strokeMiterLimit":10,"scaleX":1,"scaleY":1,"angle":0,"flipX":false,"flipY":false,"opacity":1,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"rx":0,"ry":0},{"type":"group","originX":"left","originY":"top","left":-37.77,"top":-131,"width":206.5,"height":99.67,"fill":"rgb(0,0,0)","stroke":null,"strokeWidth":0,"strokeDashArray":null,"strokeLineCap":"butt","strokeLineJoin":"miter","strokeMiterLimit":10,"scaleX":1,"scaleY":1,"angle":45,"flipX":false,"flipY":false,"opacity":1,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"objects":[{"type":"circle","originX":"left","originY":"top","left":22.25,"top":-49.83,"width":80,"height":80,"fill":"rgb(0,255,0)","stroke":null,"strokeWidth":1,"strokeDashArray":null,"strokeLineCap":"butt","strokeLineJoin":"miter","strokeMiterLimit":10,"scaleX":1,"scaleY":1,"angle":0,"flipX":false,"flipY":false,"opacity":0.5,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":10,"radius":40,"startAngle":0,"endAngle":6.283185307179586},{"type":"rect","originX":"left","originY":"top","left":-77.75,"top":-19.83,"width":50,"height":50,"fill":"rgb(255,255,0)","stroke":null,"strokeWidth":1,"strokeDashArray":null,"strokeLineCap":"butt","strokeLineJoin":"miter","strokeMiterLimit":10,"scaleX":1,"scaleY":1,"angle":30,"flipX":false,"flipY":false,"opacity":1,"shadow":null,"visible":true,"clipTo":null,"backgroundColor":"","fillRule":"nonzero","globalCompositeOperation":"source-over","transformMatrix":null,"skewX":0,"skewY":0,"rx":0,"ry":0}]}]}],"background":""}
```

**Comment by asturur** [comment]
i left some question @kangax, if you can give some insight on existing code.

File: src/canvas.class.js [hunk 1]
```diff
@@ -860,38 +858,37 @@
     /**
      * @private
      */
-    _isLastRenderedObject: function(e) {
+    _isLastRenderedObject: function(pointer) {
```
Comment (asturur) [inline src/canvas.class.js]: this function is a shortcut to check last selected target before all. But why just with controls above overaly? i think it should be checked always.

File: src/canvas.class.js [hunk 2]
```diff
@@ -940,22 +937,24 @@
...
       // Cache all targets where their bounding box contains point.
-      var target,
-          pointer = this.getPointer(e, true),
-          i = this._objects.length;
+      var target, i = objects.length, normalizedPointer;
       // Do not check for currently grouped objects, since we check the parent group itself.
       // untill we call this function specifically to search inside the activeGroup
       while (i--) {
-        if ((!this._objects[i].group || skipGroup) && this._checkTarget(e, this._objects[i], pointer)){
-          this.relatedTarget = this._objects[i];
```
Comment (asturur) [inline src/canvas.class.js]: this.relatedTarget was here, but undocumented as i can see. It gets override by canvas.targets that now will keep the list last found targets

File: src/canvas.class.js [hunk 3]
```diff
@@ -940,22 +937,24 @@
...
-          target = this._objects[i];
+        if (this._checkTarget(pointer, objects[i])) {
+          target = objects[i];
+          this.targets.push(target);
+          if (target.type === 'group') {
+            normalizedPointer = this._normalizePointer(target, pointer);
```
Comment (asturur) [inline src/canvas.class.js]: i made normalizePointer callable explicity, outside the scope of an activeGroup

File: src/mixins/canvas_events.mixin.js [hunk 4]
```diff
@@ -271,10 +271,8 @@
       if (target) {
         target.isMoving = false;
       }
-
-      shouldRender && this.renderAll();
-
       this._handleCursorAndEvent(e, target);
```
Comment (asturur) [inline src/mixins/canvas_events.mixin.js]: why mouseup is managed different from other events? it gets its own function to fire the events.

File: src/mixins/canvas_events.mixin.js [hunk 5]
```diff
@@ -352,8 +352,8 @@
       this.fire('mouse:down', { e: e });
 
       var target = this.findTarget(e);
-      if (typeof target !== 'undefined') {
-        target.fire('mousedown', { e: e, target: target });
+      for (var i = 0; i < this.targets.length; i++) {
```
Comment (asturur) [inline src/mixins/canvas_events.mixin.js]: this is the only way i could come out to manage one-click-many-targets events firing. Does it looks ok for you?

**Comment by asturur** [comment]
Also i have still to figure out mouse over and mouse out, but can be done. Also perPixelTargetFind  needs some adjustment.

**Comment by VictorHagsand** [comment]
No update on this?

**Comment by asturur** [comment]
I did not forget it.
I have to rebase on new code, but definitely pushing this forward.
It was done, just i need to check every shift click situations and transparency settings and DOCS.

**Comment by samuelhorwitz** [comment]
Hey, just a quick question. Would this update make it so that grouped objects can be manipulated individually? I have been working on a workaround for my own project in which I use a group subclass that can ungroup and regroup itself as needed so that objects can be manipulated on an individual level. However I feel like it's a bit messy. If the intent of this PR is to allow individual object manipulation, then I'll hold off on cleaning up my code and just wait for this to land so I can re-implement my stuff in a simpler fashion.

**Comment by asturur** [comment]
no just event handling, click mouse over, mouse out
.. and so on. what you say could be next step. but not this

**Comment by samuelhorwitz** [comment]
Ah okay. Well it still seems like you've done the "hard part" unless I'm missing something. Once you can reliably target objects in a group the way you've done, it's just a small step from providing an interface that allows one to "drill down" to that object and allow current transforms to get setup on that object (while making sure that the group gets updated with every change to a child object). I have a massively subclassed version of Fabric running for my project (1000s of lines of Canvas and Object subclassing) but I think that implementing this via a subclass might not be the best approach. I'm going to look into taking this branch and adding what I'm talking about on top of it to see how it feels.

File: src/canvas.class.js [hunk 6]
```diff
@@ -1293,3 +1292,4 @@
    */
   fabric.Element = fabric.Canvas;
 })();
+s
```
Comment (samuelhorwitz) [inline src/canvas.class.js]: this should probably not be here :)

**Comment by asturur** [comment]
check #2997

**Pull request closed without merging**